dev/src/bulk-writer.ts

```diff
@@ -738,29 +849,124 @@ export class BulkWriter {
    * Sends the provided batch and processes the results. After the batch is
    * committed, sends the next group of ready batches.
    *
+   * @param batchCompletedDeferred A deferred promise that resolves when the
+   * batch has been sent and received.
    * @private
    */
-  private sendBatch(batch: BulkCommitBatch): void {
-    const success = this.rateLimiter.tryMakeRequest(batch.opCount);
+  private sendBatch(
+    batch: BulkCommitBatch,
+    batchQueue: BulkCommitBatch[],
+    batchCompletedDeferred: Deferred<void>
+  ): Promise<void> {
+    const success = this._rateLimiter.tryMakeRequest(batch.opCount);
     assert(success, 'Batch should be under rate limit to be sent.');
-    batch.bulkCommit().then(() => {
+    return batch.bulkCommit().then(() => {
       // Remove the batch from the BatchQueue after it has been processed.
-      const batchIndex = this.batchQueue.indexOf(batch);
+      const batchIndex = batchQueue.indexOf(batch);
       assert(batchIndex !== -1, 'The batch should be in the BatchQueue');
-      this.batchQueue.splice(batchIndex, 1);
+      batchQueue.splice(batchIndex, 1);
+
+      batchCompletedDeferred.resolve();
+      this._pendingBatches.delete(batchCompletedDeferred.promise);
 
-      this.sendReadyBatches();
+      if (batchQueue === this._retryBatchQueue) {
+        this._retryBatchQueue.forEach(batch => batch.markReadyToSend());
```

Comment:
Who is going to send these batches?  Don't we need to schedule a commit here so that the retries work even without `flush()`? If we do so, we might not need to retry in `_flush()`.

Reply:
Changed to power all retry batch scheduling to `sendBatch()`. Now, `flush()` schedules writes on the batchQueue, then kicks off the retryBatch if there are retries. Subsequent retries are now handled when the first retry batch comes back.

Also, I moved the `markReadyToSend()` mapping for retryBatches into `sendReadyBatches()`. There's two call points (flush and sendBatch) that would require a `forEach()` call, and I got bitten several times by forgetting to mark the batches as ready to send.